### DISCUSSION TITLE
How to collect refs of all elements in list?

### DISCUSSION THREAD

**Discussion by ivan-suhorukov** [lit/lit#4787] [Q&A] [answered]
Hi! Could i somehow do something like this?
```ts
class MultyRefElement extends LitElement {
  private divRefs: Ref<HTMLDivElement>[] = []

  render() {
     return this.elements.map(element => html`<div ${ref(this.divRefs)}></div>)
  }
}
```
My goal is to collect all mapped elements in one array of refs in **ref(this.divRefs)**.. Thanks.

**Comment by yorrd** [comment]
if you know the length of the array, you could initialize the divRefs with the respective amount of `createRef`s and then access them via their index. 

wondering if it's cleaner to create another webcomponent that can be used for each element that then can hold its own ref?

**Reply by ivan-suhorukov** [reply]
Hm, i will try it, maybe it will work..

**Comment by justinfagnani** [comment]
Interesting problem...

This might be a case where it's just easier to query for the elements, either using the `@queryAll()` decorator, or querying when you need the elements.

**Reply by ivan-suhorukov** [reply]
Yes, i think i will do querying. But i suggests it will be an expensive operation, specialy in my case, when i need access to drawed children from parent quite often. I'm making a tree of nodes, and each parent node should access children. Maybe i can cache querying result and somehow detect when litelement updated to reset cache?

**Reply by justinfagnani** [reply, marked as answer]
You could cache the query and invalidate the cache in `updated()`, or even better, when the data that produces the list changes.

**Reply by justinfagnani** [reply]
Our `@query()` decorator has a cache option, but it's mean for static template elements that will never changes, since the cache never invalidates.

It'd be interesting to look at a cache option for `@queryAll()` that maybe invalidates when certain properties change.

**Reply by ivan-suhorukov** [reply]
Thanks for help!

**Comment by blikblum** [comment]
I use ref in a list of elements to initialize a JS library (Bootstrap Toast). You can look at source at: https://github.com/blikblum/luipack-js/blob/master/bs-components/toast-container.js

**Comment by dsokur** [comment]
@justinfagnani One thing is not really clear to me when @query decorator VS createRef directive should be used? Both provide the capability to get access to DOM element. As far as I understand for any static DOM elements we can use @query in other cases we need to use  createRef. Any articles of discussion will be helpful. Thanks!
